### PR TITLE
chore: migration to NextJS + materialUI | upgrade wagmi + rainbowkit + viem

### PR DESCRIPTION
**Pull request by 0xtiti**
## 

## Checklist before requesting a review

- [x] I have conducted a self-review of my code.
- [ ] If it is a core feature, I have included comprehensive tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one sentence about this update.

### CODE REVIEW

**Comment by 0xArdy** [comment]
@turtlemoji what do you think about changing `yarn` to `pnpm`? We use `pnpm` in the superchain bridge repo bc yarn was giving problems when building and running the app locally with nextjs

**Comment by turtlemoji** [comment]
> @turtlemoji what do you think about changing `yarn` to `pnpm`? We use `pnpm` in the superchain bridge repo bc yarn was giving problems when building and running the app locally with nextjs

I'm ok with pnpm, but I don't know If I like the idea to setup nextjs as default. Not all projects need it.

File: package.json [hunk 1]
```diff
@@ -1,63 +1,75 @@
...
     "@types/lodash.merge": "4.6.7",
     "@types/react": "18.2.15",
     "@types/react-dom": "18.2.7",
     "@typescript-eslint/eslint-plugin": "6.0.0",
     "@typescript-eslint/parser": "6.0.0",
-    "@vitejs/plugin-react": "4.0.3",
+    "cypress": "13.7.0",
     "eslint": "8.45.0",
     "eslint-config-prettier": "9.0.0",
+    "eslint-formatter-codeframe": "7.32.1",
```
Comment (turtlemoji) [inline package.json:61]: is this package not deprecated?
Also, idk what this package does very well. Since we are using prettier to format, the name of this package suggest that it formats code too, but we don't need any more format right?
Is a dependency for next?

File: cypress/support/commands.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/// <reference types="cypress" />
```
Comment (turtlemoji) [inline cypress/support/commands.ts:1]: we should keep this?

File: .env.example [hunk 3]
```diff
@@ -1,3 +1,3 @@
-VITE_RPC_URL= # Example: https://localhost:8545
-VITE_PROJECT_ID= # ProjectId from walletConnect
-VITE_ALCHEMY_KEY= # API key from Alchemy 
\ No newline at end of file
+NEXT_PUBLIC_RPC_URL= # Example: https://localhost:8545
+NEXT_PUBLIC_PROJECT_ID= # API from WalletConnect
```
Comment (turtlemoji) [inline .env.example:2]: is this an api endpoint or a project id then?

File: cypress/e2e/fundamentals.cy.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+describe('template spec', () => {
+  it('passes', () => {
+    cy.visit('https://example.cypress.io');
```
Comment (turtlemoji) [inline cypress/e2e/fundamentals.cy.ts:3]: change default examples, maybe

File: cypress/fixtures/example.json [hunk 5]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (turtlemoji) [inline cypress/fixtures/example.json:1]: change or delete examples

File: package.json [hunk 6]
```diff
@@ -1,63 +1,75 @@
...
     "lint": "eslint 'src/**/*.{js,jsx,ts,tsx}'",
     "lint:fix": "yarn run lint -- --fix",
     "prettier": "prettier src --check",
     "prettier:fix": "yarn run prettier -- --write",
     "format": "yarn run prettier:fix && yarn run lint:fix",
     "format:check": "yarn run prettier && yarn run lint",
     "prepare": "husky install && wonderland-crypto-husky-checks install",
-    "test": "vitest run --dir src/tests/",
-    "test:watch": "vitest --dir src/tests/"
+    "test": "yarn cypress run",
```
Comment (turtlemoji) [inline package.json:23]: add unit testing

File: src/config/env.ts [hunk 7]
```diff
@@ -1,11 +1,11 @@
...
 
 export const getEnv = (): Env => {
-  const { VITE_RPC_URL, VITE_PROJECT_ID, VITE_ALCHEMY_KEY } = import.meta.env;
+  const { NEXT_PUBLIC_RPC_URL, NEXT_PUBLIC_PROJECT_ID, NEXT_PUBLIC_ALCHEMY_KEY } = process.env;
 
   return {
-    RPC_URL: VITE_RPC_URL,
-    PROJECT_ID: VITE_PROJECT_ID,
-    ALCHEMY_KEY: VITE_ALCHEMY_KEY,
+    RPC_URL: NEXT_PUBLIC_RPC_URL || '',
```
Comment (turtlemoji) [inline src/config/env.ts:7]: @0xArdy @0xtiti  maybe it is better to leave it as undefined and erroring in this case

File: src/containers/Footer/index.tsx [hunk 8]
```diff
@@ -12,29 +13,29 @@ export const Footer = () => {
...
+  const { currentTheme } = useCustomTheme();
+  return {
+    display: 'flex',
+    height: '8rem',
+    padding: '0 8rem',
+    alignItems: 'center',
+    justifyContent: 'space-between',
+    backgroundColor: currentTheme.backgroundSecondary,
+    borderTop: currentTheme.border,
+    width: '100vw',
```
Comment (turtlemoji) [inline src/containers/Footer/index.tsx:26]: recommend % units instead of viewport units on child elements that are not body or html tho.

File: src/containers/Header/index.tsx [hunk 9]
```diff
@@ -1,48 +1,40 @@
...
+  const { currentTheme } = useCustomTheme();
+  return {
+    display: 'flex',
+    height: '8rem',
+    padding: '0 8rem',
+    alignItems: 'center',
+    justifyContent: 'space-between',
+    backgroundColor: currentTheme.backgroundSecondary,
+    width: '100vw',
+    zIndex: 100,
```
Comment (turtlemoji) [inline src/containers/Header/index.tsx:32]: put on constants on another pr <3 / todo

File: src/containers/Landing/index.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    <LandingContainer>
+      <h1>Web3 React Boilerplate</h1>
+    </LandingContainer>
+  );
+};
+
+const LandingContainer = styled('div')({
+  display: 'flex',
+  flexDirection: 'column',
+  height: 'calc(100vh - 16rem)',
```
Comment (turtlemoji) [inline src/containers/Landing/index.tsx:14]: don't like the idea of hardcoding this here, would be nice to rework this on another pr

File: src/pages/layout.tsx [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+  flex-direction: column;
+  align-items: center;
+  justify-content: start;
+  width: 100%;
+  max-width: 100%;
+  overflow-x: hidden;
+
+  max-width: 120rem;
+  padding: 0 4rem;
+  min-height: 100vh;
```
Comment (turtlemoji) [inline src/pages/layout.tsx:38]: 👀 maybe /todo for another pr to rework

File: src/providers/Web3ModalProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+import { config } from '../utils';
+
+type Props = {
+  children: ReactNode;
+};
+
+const queryClient = new QueryClient();
+
+export function Web3ModalProvider({ children }: Props) {
```
Comment (turtlemoji) [inline src/providers/Web3ModalProvider.tsx:16]: are we using web3modal or it is legacy?

File: src/pages/_document.tsx [hunk 13]
```diff
@@ -0,0 +1,40 @@
+import { DocumentHeadTags, documentGetInitialProps } from '@mui/material-nextjs/v13-pagesRouter';
+import type { DocumentHeadTagsProps } from '@mui/material-nextjs/v13-pagesRouter';
+import { DocumentContext, DocumentProps, Head, Html, Main, NextScript } from 'next/document';
+
+export default function MyDocument(props: DocumentProps & DocumentHeadTagsProps) {
```
Comment (turtlemoji) [inline src/pages/_document.tsx:5]: wrong `&` ?

File: src/hooks/ScrollToTop.tsx [hunk 14]
```diff
@@ -1,12 +1,19 @@
...
   useEffect(() => {
-    window.scrollTo(0, 0);
-  }, [pathname]);
+    const handleRouteChange = () => {
+      window.scrollTo(0, 0);
+    };
+    router.events.on('routeChangeComplete', handleRouteChange);
+
+    return () => {
+      router.events.off('routeChangeComplete', handleRouteChange);
```
Comment (turtlemoji) [inline src/hooks/ScrollToTop.tsx:14]: recheck for memory leaks / this is working as expected

File: src/pages/_document.tsx (hunk 13 above)
Comment (0xtiti) [reply src/pages/_document.tsx:5]: Screen Shot 2024-03-27 at 17 52 33

Comment (turtlemoji) [reply src/pages/_document.tsx:5]: wow it breaks the gh styling

File: package.json (hunk 1 above)
Comment (0xtiti) [reply package.json:61]: removed, not needed https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c

File: cypress/support/commands.ts (hunk 2 above)
Comment (0xtiti) [reply cypress/support/commands.ts:1]: https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/13af02d9fb3fdcdcf75ddd7887f7af7b9c690fee

File: .env.example (hunk 3 above)
Comment (0xtiti) [reply .env.example:2]: projectID, my bad...https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c

File: package.json (hunk 6 above)
Comment (0xtiti) [reply package.json:23]: will create a new pr for unit testing

File: src/config/env.ts (hunk 7 above)
Comment (0xtiti) [reply src/config/env.ts:7]: fixed with erroring https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c But its causing issues on github workflow build

File: src/containers/Footer/index.tsx (hunk 8 above)
Comment (0xtiti) [reply src/containers/Footer/index.tsx:26]: https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c

File: src/containers/Landing/index.tsx (hunk 10 above)
Comment (0xtiti) [reply src/containers/Landing/index.tsx:14]: https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/f97a49e8d1b84e01dea3db2ef87ddd6541413e2e

File: src/containers/Header/index.tsx (hunk 9 above)
Comment (0xtiti) [reply src/containers/Header/index.tsx:32]: https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c#diff-31f5dea1a137fef36b405b82a17c26ff39f1991fec1f17f2693bd4e7d3ed9b13

File: src/pages/layout.tsx (hunk 11 above)
Comment (0xtiti) [reply src/pages/layout.tsx:38]: https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c#diff-31f5dea1a137fef36b405b82a17c26ff39f1991fec1f17f2693bd4e7d3ed9b13

File: src/providers/Web3ModalProvider.tsx (hunk 12 above)
Comment (0xtiti) [reply src/providers/Web3ModalProvider.tsx:16]: legacy https://github.com/defi-wonderland/web3-react-boilerplate/pull/8/commits/a4a80bcb3d05cb213b045a2f40932f8ae160615c
